models/room_address.go

```diff
@@ -8,48 +8,199 @@
-	ipv6KubernetesLabelKey string
+	logger                  logrus.FieldLogger
+	useCache                bool
+	cacheExpirationInterval time.Duration
+	ipv6KubernetesLabelKey  string
 }
 
 // NewRoomAddressesFromHostPort is the RoomAddressesFromHostPort constructor
-func NewRoomAddressesFromHostPort(ipv6KubernetesLabelKey string) *RoomAddressesFromHostPort {
+func NewRoomAddressesFromHostPort(
+	logger logrus.FieldLogger, ipv6KubernetesLabelKey string,
+	useCache bool, cacheExpirationInterval time.Duration,
+) *RoomAddressesFromHostPort {
 	return &RoomAddressesFromHostPort{
-		ipv6KubernetesLabelKey: ipv6KubernetesLabelKey,
+		ipv6KubernetesLabelKey:  ipv6KubernetesLabelKey,
+		useCache:                useCache,
+		cacheExpirationInterval: cacheExpirationInterval,
+		logger:                  logger,
 	}
 }
 
 // Get gets room public addresses
-func (r *RoomAddressesFromHostPort) Get(room *Room, kubernetesClient kubernetes.Interface) (*RoomAddresses, error) {
-	return getRoomAddresses(false, r.ipv6KubernetesLabelKey, room, kubernetesClient)
+func (r *RoomAddressesFromHostPort) Get(
+	room *Room, kubernetesClient kubernetes.Interface, redisClient redisinterfaces.RedisClient,
+) (*RoomAddresses, error) {
+	if addrs := r.fromCache(redisClient, room); addrs != nil {
+		return addrs, nil
+	}
+	addrs, err := getRoomAddresses(false, r.ipv6KubernetesLabelKey, room, kubernetesClient)
+	if err != nil {
+		return nil, err
+	}
+	r.toCache(redisClient, room, addrs)
+	return addrs, nil
+}
+
+func (r RoomAddressesFromHostPort) toCache(
+	redisClient redisinterfaces.RedisClient, room *Room, addrs *RoomAddresses,
+) {
+	if !r.useCache {
+		return
+	}
+	l := r.logger.WithFields(logrus.Fields{
+		"struct": "RoomAddressesFromNodePort",
+		"method": "toCache",
+	})
+	b, err := json.Marshal(addrs)
+	if err != nil {
+		l.WithError(err).Error("failed to marshal room addresses")
+		return
+	}
+	if redisClient.Set(r.buildCacheKey(room), b, r.cacheExpirationInterval).Err() != nil {
+		l.WithError(err).Error("failed to set room addresses in redis")
+	}
+}
+
+func (r RoomAddressesFromHostPort) fromCache(redisClient redisinterfaces.RedisClient, room *Room) *RoomAddresses {
+	if !r.useCache {
+		return nil
+	}
+	cmd := redisClient.Get(r.buildCacheKey(room))
+	err := cmd.Err()
+	l := r.logger.WithFields(logrus.Fields{
+		"struct": "RoomAddressesFromNodePort",
+		"method": "fromCache",
+	})
+	if err != nil && err != redis.Nil {
+		l.WithError(err).Error("failed to get from redis")
+		return nil
+	}
+	if err == redis.Nil {
+		return nil
+	}
+	addrs := &RoomAddresses{}
+	b, err := cmd.Bytes()
+	if err != nil {
+		l.WithError(err).Error("failed to cmd.Bytes()")
+		return nil
+	}
+	if err := json.Unmarshal(b, addrs); err != nil {
+		l.WithError(err).Error("failed unmarshal room addresses from redis")
+		return nil
+	}
+	return addrs
+}
+
+func (r *RoomAddressesFromHostPort) buildCacheKey(room *Room) string {
+	return fmt.Sprintf("%s-%s", room.SchedulerName, room.ID)
```

Comment:
I'm pretty sure this could conflict with other keys. Maybe something a bit more specific like
`fmt.Sprintf("addr-%s-%s", room.SchedulerName, room.ID) `

Reply:
makes complete sense, will do